From fc81ac5b4046b097ef5bdb2fa8820eff7e0262b7 Mon Sep 17 00:00:00 2001
From: Alexandre Dumont <adumont@gmail.com>
Date: Thu, 25 Oct 2012 22:47:32 +0200
Subject: [PATCH 1/2] Change proximity behaviour for devices with fixed
 keyboard

The actual behaviour of Phone (ignore proximity sensor and force screen to stay
on when the keyboard is opened) while correct for a slidding keyboard, is not
correct in the case of a device with a fixed opened keyboard, like the HTC Chacha.

With this patch, you can indicate in the overlay that the device has a fixed
keyboard. When done so, the state of the keyboard will be ignored, and the screen
will be turned on/off based on the proximity sensor.

Change-Id: If6b06cd8df3173e1bddec02a87892ef6e71dcbca
---
 res/values/config.xml               |  4 ++++
 src/com/android/phone/PhoneApp.java | 10 ++++++++--
 2 files changed, 12 insertions(+), 2 deletions(-)

diff --git a/res/values/config.xml b/res/values/config.xml
index 5b671e1..1814773 100644
--- a/res/values/config.xml
+++ b/res/values/config.xml
@@ -122,4 +122,8 @@
          for the "Phantom Voicemail" issue where notifications that are
          impossible to dismiss on the Sprint network. -->
     <bool name="sprint_mwi_quirk">false</bool>
+
+    <!-- True if the device has a fixed keyboard (as oposite to a sliding one) so keyboard is
+	always open. Will modify proximity checking behaviour. -->
+    <bool name="config_device_has_fixed_keyboard">false</bool>
 </resources>
diff --git a/src/com/android/phone/PhoneApp.java b/src/com/android/phone/PhoneApp.java
index e7bc04e..f48a1d9 100644
--- a/src/com/android/phone/PhoneApp.java
+++ b/src/com/android/phone/PhoneApp.java
@@ -217,6 +217,9 @@ public class PhoneApp extends Application implements AccelerometerListener.Orien
 
     private WakeState mWakeState = WakeState.SLEEP;
 
+    // True if the keyboard is fixed (not sliding). Defined in device overlay
+    private boolean mIsKeyboardAlwaysOpen;
+
     /**
      * Timeout setting used by PokeLock.
      *
@@ -578,6 +581,9 @@ public class PhoneApp extends Application implements AccelerometerListener.Orien
             // Read platform settings for TTY feature
             mTtyEnabled = getResources().getBoolean(R.bool.tty_enabled);
 
+            // Read platform setting regarding keyboard
+            mIsKeyboardAlwaysOpen = getResources().getBoolean(R.bool.config_device_has_fixed_keyboard);
+
             // Register for misc other intent broadcasts.
             IntentFilter intentFilter =
                     new IntentFilter(Intent.ACTION_AIRPLANE_MODE_CHANGED);
@@ -1310,13 +1316,13 @@ public class PhoneApp extends Application implements AccelerometerListener.Orien
             synchronized (mProximityWakeLock) {
                 // turn proximity sensor off and turn screen on immediately if
                 // we are using a headset and is not configured to keep sensor on
-                // the keyboard is open, or the device
+                // the keyboard is open (unless it is a fixed keyboard), or the device
                 // is being held in a horizontal position.
                 boolean keepOn = PhoneUtils.PhoneSettings.keepProximitySensorOn(this);
                 boolean screenOnImmediately = ((!keepOn && isHeadsetPlugged())
                             || PhoneUtils.isSpeakerOn(this)
                             || (!keepOn && (mBtHandsfree != null) && mBtHandsfree.isAudioOn())
-                            || mIsHardKeyboardOpen);
+                            || (mIsHardKeyboardOpen  && !mIsKeyboardAlwaysOpen));
 
                 // We do not keep the screen off when the user is outside in-call screen and we are
                 // horizontal, but we do not force it on when we become horizontal until the
-- 
1.8.0

